functional-tests/appwrapper_reconciler_test.go

```diff
@@ -0,0 +1,232 @@
+
+	t.Cleanup(func() {
+		teardownTestEnv(t, testEnv)
+	})
+
+	return testEnv, cfg
+}
+
+func startInstascaleController(test Test, testEnv *envtest.Environment, cfg *rest.Config) {
+	// Add custom resource schemes to the global scheme.
+	err := mcadv1beta1.AddToScheme(scheme.Scheme)
+	test.Expect(err).NotTo(HaveOccurred())
+
+	err = machinev1beta1.AddToScheme(scheme.Scheme)
+	test.Expect(err).NotTo(HaveOccurred())
+
+	// Create a controller manager for managing controllers.
+	k8sManager, err := ctrl.NewManager(cfg, ctrl.Options{
+		Scheme: scheme.Scheme,
+	})
+	test.Expect(err).ToNot(HaveOccurred())
+
+	// Create an instance of the AppWrapperReconciler with configuration
+	instaScaleController := &controllers.AppWrapperReconciler{
+		Client: k8sManager.GetClient(),
+		Scheme: k8sManager.GetScheme(),
+		Config: config.InstaScaleConfiguration{
+			MachineSetsStrategy: "reuse",
+			MaxScaleoutAllowed:  5,
+		},
+	}
+	// Set up the AppWrapperReconciler with the manager.
+	err = instaScaleController.SetupWithManager(context.Background(), k8sManager)
+	test.Expect(err).ToNot(HaveOccurred())
+
+	// Start the controller manager in a goroutine.
+	go func() {
+		err = k8sManager.Start(ctrl.SetupSignalHandler())
+		test.Expect(err).ToNot(HaveOccurred())
+	}()
+
+}
+
+func teardownTestEnv(t *testing.T, testEnv *envtest.Environment) {
+	if err := testEnv.Stop(); err != nil {
+		t.Log("Error stopping test Environment\n", err)
+	}
+}
+
+func instascaleAppwrapper(namespace string) *mcadv1beta1.AppWrapper {
+	aw := &mcadv1beta1.AppWrapper{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-instascale",
+			Namespace: namespace,
+			Labels: map[string]string{
+				"orderedinstance": "test.instance1",
+			},
+			Finalizers: []string{"instascale.codeflare.dev/finalizer"},
+		},
+		Spec: mcadv1beta1.AppWrapperSpec{
+			AggrResources: mcadv1beta1.AppWrapperResourceList{
+				GenericItems: []mcadv1beta1.AppWrapperGenericResource{
+					{
+						DesiredAvailable: 1,
+						CustomPodResources: []mcadv1beta1.CustomPodResourceTemplate{
+							{
+								Replicas: 1,
+								Requests: apiv1.ResourceList{
+									apiv1.ResourceCPU:    resource.MustParse("250m"),
+									apiv1.ResourceMemory: resource.MustParse("512Mi"),
+								},
+								Limits: apiv1.ResourceList{
+									apiv1.ResourceCPU:    resource.MustParse("1"),
+									apiv1.ResourceMemory: resource.MustParse("1G"),
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	return aw
+}
+
+func MachineSet(test Test, namespace string, machineSetName string) func(g Gomega) *machinev1beta1.MachineSet {
+	return func(g Gomega) *machinev1beta1.MachineSet {
+		machineset, err := test.Client().Machine().MachineV1beta1().MachineSets(namespace).Get(test.Ctx(), machineSetName, metav1.GetOptions{})
+		test.Expect(err).ToNot(HaveOccurred())
+		return machineset
+	}
+}
+
+func MachineSetReplicas(machineSet *machinev1beta1.MachineSet) *int32 {
+	return machineSet.Spec.Replicas
+}
```

Comment:
These functions can be moved to https://github.com/project-codeflare/codeflare-common/blob/main/support/machine.go